File: CS/T591683/MainForm.cs
using DevExpress.Utils.Filtering;
using DevExpress.Utils.Filtering.Internal;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.Filtering;
using System;
using System.Data;
using System.Linq;

namespace T591683
{
    public partial class MainForm : DevExpress.XtraEditors.XtraForm
    {
        ICollectionValueViewModel<string> viewModel;
        CheckedListBoxItem[] items;
        CheckedListBoxItem[] savedItems;
        string columnName = String.Empty;

        public MainForm()
        {
            InitializeComponent();
        }

        private void MainForm_Load(object sender, EventArgs e)
        {
            SampleDataGenerator dataGenerator = new SampleDataGenerator();
            gridControl1.DataSource = dataGenerator.CreateTable();
            gridView1.FilterPopupExcelPrepareTemplate += GridView1_FilterPopupExcelPrepareTemplate;
            gridView1.FilterPopupExcelCustomizeTemplate += GridView1_FilterPopupExcelCustomizeTemplate;
        }
        private void GridView1_FilterPopupExcelPrepareTemplate(object sender, DevExpress.XtraGrid.Views.Grid.FilterPopupExcelPrepareTemplateEventArgs e)
        {
            e.Template = new CustomTemplate();
        }

        private void GridView1_FilterPopupExcelCustomizeTemplate(object sender, DevExpress.XtraGrid.Views.Grid.FilterPopupExcelCustomizeTemplateEventArgs e)
        {
            viewModel = ((e.Template.Parent as FilterUIEditorContainerEdit).EditValue as ICollectionValueViewModel<string>);
            var checkedListBoxControl = e.GetValuesListBox();
            var searchControl = e.GetSearchControl();
            searchControl.Client = checkedListBoxControl;
            var textBox = e.GetTemplatedChild<SpinEdit>("Part_StringLength");
            savedItems = checkedListBoxControl.Items.ToArray();

            textBox.EditValueChanged += (s, te) =>
            {
                if (e.Column.FieldName != columnName)
                {
                    items = checkedListBoxControl.Items.ToArray();
                    columnName = e.Column.FieldName;
                }

                checkedListBoxControl.BeginUpdate();
                checkedListBoxControl.Items.Clear();
                for (int i = 0; i < items.Length; i++) 
                    if (i == 0 || items[i].Description.Length >= textBox.Value)
                        checkedListBoxControl.Items.Add(items[i]);
                checkedListBoxControl.EndUpdate();
            };
        }
    }
}

File: VB/T591683/MainForm.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Utils.Filtering
Imports DevExpress.Utils.Filtering.Internal
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.Filtering
Imports System
Imports System.Data
Imports System.Linq

Namespace T591683
	Partial Public Class MainForm
		Inherits DevExpress.XtraEditors.XtraForm
		Private viewModel As ICollectionValueViewModel(Of String)
		Private items() As CheckedListBoxItem
		Private savedItems() As CheckedListBoxItem
		Private columnName As String = String.Empty

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub MainForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			Dim dataGenerator As New SampleDataGenerator()
			gridControl1.DataSource = dataGenerator.CreateTable()
			AddHandler gridView1.FilterPopupExcelPrepareTemplate, AddressOf GridView1_FilterPopupExcelPrepareTemplate
			AddHandler gridView1.FilterPopupExcelCustomizeTemplate, AddressOf GridView1_FilterPopupExcelCustomizeTemplate
		End Sub
		Private Sub GridView1_FilterPopupExcelPrepareTemplate(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Grid.FilterPopupExcelPrepareTemplateEventArgs)
			e.Template = New CustomTemplate()
		End Sub

		Private Sub GridView1_FilterPopupExcelCustomizeTemplate(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Grid.FilterPopupExcelCustomizeTemplateEventArgs)
			viewModel = (TryCast((TryCast(e.Template.Parent, FilterUIEditorContainerEdit)).EditValue, ICollectionValueViewModel(Of String)))
			Dim checkedListBoxControl As CheckedListBoxControl = e.GetValuesListBox()
			Dim searchControl As SearchControl = e.GetSearchControl()
			searchControl.Client = checkedListBoxControl
			Dim spinEditControl As SpinEdit = e.GetTemplatedChild(Of SpinEdit)("Part_StringLength")
			savedItems = checkedListBoxControl.Items.ToArray()

            AddHandler spinEditControl.EditValueChanged, Function(s, te) AnonymousMethod1(s, te, e, checkedListBoxControl, spinEditControl)
        End Sub

        Private Function AnonymousMethod1(ByVal s As Object, ByVal te As Object, ByVal e As DevExpress.XtraGrid.Views.Grid.FilterPopupExcelCustomizeTemplateEventArgs, ByVal checkedListBoxControl As CheckedListBoxControl, ByVal spinEditControl As SpinEdit) As Boolean
            If e.Column.FieldName <> columnName Then
                items = checkedListBoxControl.Items.ToArray()
                columnName = e.Column.FieldName
            End If
            checkedListBoxControl.BeginUpdate()
            checkedListBoxControl.Items.Clear()
            For i As Integer = 0 To items.Length - 1
                If i = 0 OrElse items(i).Description.Length >= spinEditControl.Value Then
                    checkedListBoxControl.Items.Add(items(i))
                End If
            Next i
            checkedListBoxControl.EndUpdate()
            Return True
        End Function
    End Class
End Namespace
